contracts/GenericSwap.sol

```diff
@@ -0,0 +1,108 @@
+import { EIP712 } from "./abstracts/EIP712.sol";
+import { IGenericSwap } from "./interfaces/IGenericSwap.sol";
+import { IStrategy } from "./interfaces/IStrategy.sol";
+import { GeneralAsset } from "./libraries/GeneralAsset.sol";
+import { SignatureValidator } from "./libraries/SignatureValidator.sol";
+
+contract GenericSwap is IGenericSwap, TokenCollector, EIP712 {
+    using SafeERC20 for IERC20;
+    using GeneralAsset for address;
+
+    bytes32 public constant GS_DATA_TYPEHASH = 0x9a6d9096b182513baa520ad9d0766c6e70d0637fcf40521146c04435396a0fdf;
+
+    mapping(bytes32 => bool) private filledSwap;
+
+    /*
+        keccak256(
+            abi.encodePacked(
+                "GenericSwapData(",
+                "address inputToken,",
+                "address outputToken,",
+                "uint256 inputAmount,",
+                "uint256 minOutputAmount,",
+                "address receiver,",
+                "uint256 deadline,",
+                "bytes inputData,",
+                "bytes strategyData,",
+                "uint256 salt",
+                ")"
+            )
+        );
+        */
+
+    constructor(address _uniswapPermit2) TokenCollector(_uniswapPermit2) {}
+
+    /// @param strategy The strategy contract
+    /// @param swapData Swap data
+    /// @return returnAmount Output amount of the swap
+    function executeSwap(IStrategy strategy, GenericSwapData calldata swapData) external payable override returns (uint256 returnAmount) {
+        return _executeSwap(strategy, swapData, msg.sender);
+    }
+
+    /// @param strategy The strategy contract
+    /// @param swapData Swap data
+    /// @return returnAmount Output amount of the swap
+    function executeSwap(
+        IStrategy strategy,
+        GenericSwapData calldata swapData,
+        address taker,
+        uint256 salt,
+        bytes calldata takerSig
+    ) external payable override returns (uint256 returnAmount) {
+        bytes32 swapHash = getEIP712Hash(_getGSDataHash(swapData, salt));
+        require(!filledSwap[swapHash], "already filled");
+        if (!SignatureValidator.isValidSignature(taker, swapHash, takerSig)) revert InvalidSignature();
+        filledSwap[swapHash] = true;
+        return _executeSwap(strategy, swapData, taker);
+    }
+
+    function _executeSwap(
+        IStrategy strategy,
+        GenericSwapData calldata swapData,
+        address taker
+    ) private returns (uint256 returnAmount) {
+        address _inputToken = swapData.inputToken;
+        address _outputToken = swapData.outputToken;
+
+        if (_inputToken.isETH() && msg.value != swapData.inputAmount) revert InvalidMsgValue();
+
+        if (!_inputToken.isETH()) {
+            _collect(_inputToken, taker, address(strategy), swapData.inputAmount, swapData.inputData);
+        }
+
+        strategy.executeStrategy{ value: msg.value }(_inputToken, _outputToken, swapData.inputAmount, swapData.strategyData);
```

Comment:
Should wrap ETH in strategy  contract?

Reply:
It's left for strategy contract handling this since for some amm protocols, raw ETH is supported.